torch/sparse/semi_structured/semi_structured_sparse_tensor.py

```diff
@@ -0,0 +1,156 @@
+import torch
+
+
+class SemiStructuredSparseTensor(torch.Tensor):
+    """
+    This class implementes semi-structured (2:4) sparsity as a Tensor subclass.
+
+    Semi-structured sparsity describes a sparsity pattern where n in every 2n elements are sparse,
+    depending on the datatype. It is most commonly referred to as 2:4 sparsity or fine-grained
+    structured sparsity.
+
+    compressed_tensor is a tensor that stores both the kept elemenets and the metadata mask
+    These two are stored next to each other in one contiguous tensor.
+
+    compressed tensor = [ kept elemetns of original tensor |   mask_metadata     ]
+
+    For an original tensor of size (m, k) we expect the first m * k // 2 elements to be the kept elements
+    The rest of the tensor is metadata.
+
+    This tensor subclass also has a cslt object, which is set if pytorch is built with cuSPARSELt support.
+    In this case, we store some additional metadata containg the sparse matrix descriptor in order for
+    faster matmul performance.
+
+    When this cslt object is not set, we will default to using CUTLASS kernels and _structured_sparse_linear.
+    """
+
+    # When fuse_transpose is set to True, we fuse a .T into the cuSPASRELt matmul operation.
+    # We do this in a bit of a hacky manner, by just creating a transposed matrix and then setting
+    # the output order to be Column major instead of Row major.
+    fuse_transpose = True
+
+    @staticmethod
+    def __new__(cls, custom_shape, compressed_tensor, cslt, transposed):
+        kwargs = {}
+        kwargs["device"] = compressed_tensor.device
+        kwargs["dtype"] = compressed_tensor.dtype
+        kwargs["layout"] = compressed_tensor.layout
+        kwargs["requires_grad"] = False
+
+        return torch.Tensor._make_wrapper_subclass(cls, custom_shape, **kwargs)  # type: ignore[attr-defined]
+
+    def __init__(
+        self,
+        custom_shape,
+        compressed_tensor: torch.Tensor,
+        cslt,
+        transposed,
+    ):
+        self.compressed_tensor = compressed_tensor
+        self.cslt = cslt
+        self.transposed = transposed
+
+    @property
+    def kept_elements(self):
+        m, k = self.shape
+        num_kept_elements = m * k // 2
+        return self.compressed_tensor[:num_kept_elements].view(m, k // 2)
+
+    @property
+    def metadata(self):
```

Comment:
Aside from testing, I'm not sure what the value of this is to most users. If we want to keep it around for testing only let's add a leading underscore to make sure it's marked private.

Reply:
I think it'd be nice to have some way to view the metadata separately. 

Would be useful for going from the compressed -> dense representation, and I think users will want to see the mask. 

This also for future composability with CUTLASS, since it expects the metadata to be passed in separately. I'm thinking of something like:

```
sA = SemiStructuredSparseTensor(A)
output = _structured_sparse_linear(sA.values, sA.metadata)
```

Reply:
I think that makes sense, but conversion functions and integration of other matmul backends don't need the metadata to be publicly visible.

Reply:
BTW, I changed this so you now access the metadata by calling `.indices()`.